### PR TITLE
wifi-subsys: async uart implementation

### LINKED ISSUE #116: add uart implentation for wifi-subsys
#### Description

Add the uart implementation this will be thee protocol used for the communication, this module must have un handler for AT commands for handle the wifi and other settings also must identify the datos arrived since sensors.

### PR DESCRIPTION
**Pull request by xkevin190**
### Contribution description

simple async uart implementation for use with AT commands and could receive sensor datos for send to the cloud.

This is an initial implementation for the communication between the boards throw uart protocol.

### testing procedure

`cd test && idf.py build -D "TEST_COMPONENTS=uart" flash monitor`

```sh
Here's the test menu, pick your combo:
(1)	"uart initialize" [UART]
(2)	"send uart bytes" [UART]
```
this PR fixes #116

### CODE REVIEW

File: wifi-subsys/components/uart/src/subsys_uart.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ *
+ * @brief       uart module
+ * @ingroup     wifisubsys
```
Comment (luisan00) [inline wifi-subsys/components/uart/src/subsys_uart.h:20]:
```suggestion
 * @ingroup     wifi-subsys
```

File: wifi-subsys/components/uart/src/subsys_uart.c [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+        value = (char *) malloc(strlen(key2) +1);
+        memcpy(value, key2, strlen(key2) +1);
+      }
+      i++;
+      key2 = strtok(NULL, "+");
+    }
+    /* cppcheck-suppress nullPointerRedundantCheck
+    * (reason: <your reason why you think this is a false positive>) */
+    memcpy(&output->key, key_value, strlen(key_value) +1);
+    /* cppcheck-suppress nullPointerRedundantCheck
```
Comment (luisan00) [inline wifi-subsys/components/uart/src/subsys_uart.c:103]: And this lines are for?

File: wifi-subsys/components/uart/src/subsys_uart.c [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+        }
+    }
+
+    return -1;
+}
+
+void at_handler(uint8_t* at_comant)
+{
+    char* commands = (char*)at_comant;
+    // ESP_LOGI(TAG, "Read bytes: '%s' ", commands);
```
Comment (luisan00) [inline wifi-subsys/components/uart/src/subsys_uart.c:134]: Remove dead code

File: wifi-subsys/components/uart/src/subsys_uart.c [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    while (1) {
+        const int rxBytes = uart_read_bytes(UART_NUM_1, data, RX_BUF_SIZE, 1000 / portTICK_RATE_MS);
+        if (rxBytes > 0) {
+            data[rxBytes] = 0;
+            if(rxBytes > 2 && data[0] == 0x41 && data[1] == 0x54 && data[2] == 0x2b){
+                at_handler(data);
+            } else {
+                received_sensor_data(data);
+            }
+            // ESP_LOG_BUFFER_HEXDUMP(TAG, data, rxBytes, ESP_LOG_INFO);
```
Comment (luisan00) [inline wifi-subsys/components/uart/src/subsys_uart.c:219]: Dead code ?

File: wifi-subsys/components/uart/src/subsys_uart.h [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ *
```
Comment (luisan00) [inline wifi-subsys/components/uart/src/subsys_uart.h:18]:
```suggestion
```

File: wifi-subsys/components/uart/src/subsys_uart.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+extern "C"
+{
+#endif
+/**
+ * @brief this struct content the key and value of the AT request
+ *
+ */
+typedef struct {
+    char value[50]; /*!< AT value */
+    char key[50];   /*!< AT key */
```
Comment (luisan00) [inline wifi-subsys/components/uart/src/subsys_uart.h:41]: Which's the longest key?

Comment (luisan00) [reply wifi-subsys/components/uart/src/subsys_uart.h:41]: Anyway we shouldn't use static length arrays

**Review by luisan00 — CHANGES_REQUESTED** [review]
Why main.c is being updated ? XD

File: wifi-subsys/components/uart/src/subsys_uart.h (hunk 6 above)
Comment (xkevin190) [reply wifi-subsys/components/uart/src/subsys_uart.h:41]: seems me a reasonable size for the keys and value

Comment (luisan00) [reply wifi-subsys/components/uart/src/subsys_uart.h:40]: But which is the longest key?

**Comment by luisan00** [comment]
Looks good for me,  do a squash and I think we are ready to move it forward

**Review by luisan00 — APPROVED** [review]
Good job man, 👏

**Pull request merged**